src/main/java/org/spongepowered/api/command/args/GenericArguments.java

```diff
@@ -1585,4 +1604,277 @@ public Text getUsage(CommandSource src) {
+     */
+    public static CommandElement url(Text key) {
+        return new URIElement(key, false);
+    }
+
+    /**
+     * Expect an argument to represent a {@link URI}.
+     *
+     * @param key The key to store under
+     * @return the argument
+     */
+    public static CommandElement uri(Text key) {
+        return new URIElement(key, true);
+    }
+
+    private static class URIElement extends KeyElement {
+
+        private final boolean returnURI;
+
+        protected URIElement(Text key, boolean returnURI) {
+            super(key);
+            this.returnURI = returnURI;
+        }
+
+        @Nullable
+        @Override
+        protected Object parseValue(CommandSource source, CommandArgs args) throws ArgumentParseException {
+            String str = args.next();
+            URL url;
+            try {
+                url = new URL(str);
+            } catch (MalformedURLException ex) {
+                throw new ArgumentParseException(Text.of("Invalid URL!"), ex, str, 0);
+            }
+            URI uri;
+            try {
+                uri = url.toURI();
+            } catch (URISyntaxException ex) {
+                throw new ArgumentParseException(Text.of("Invalid URL!"), ex, str, 0);
+            }
+            if (this.returnURI) {
+                return uri;
+            } else {
+                return url;
+            }
+        }
+    }
+
+    /**
+     * Expect an argument to return an IP address, in the form of an
+     * {@link InetAddress}.
+     *
+     * @param key The key to store under
+     * @return the argument
+     */
+    public static CommandElement ip(Text key) {
+        return new IpElement(key, false);
+    }
+
+    /**
+     * Expect an argument to return an IP address, in the form of an
+     * {@link InetAddress}, or if nothing matches and the source is a
+     * {@link RemoteSource}, return the source's address.
+     *
+     * @param key The key to store under
+     * @return the argument
+     */
+    public static CommandElement ipOrSource(Text key) {
+        return new IpElement(key, true);
+    }
+
+    private static class IpElement extends KeyElement {
+
+        private final boolean self;
+        private final PlayerCommandElement possiblePlayer;
+
+        protected IpElement(Text key, boolean self) {
+            super(key);
+            this.self = self;
+            this.possiblePlayer = new PlayerCommandElement(key, false);
+        }
+
+        @Nullable
+        @Override
+        protected Object parseValue(CommandSource source, CommandArgs args) throws ArgumentParseException {
+            if (!args.hasNext() && this.self) {
+                if (source instanceof Player) {
+                    return ((Player) source).getConnection().getAddress().getAddress();
+                } else {
+                    throw args.createError(Text.of("No IP address was specified!"));
+                }
+            }
+            Object state = args.getState();
```

Comment:
That is in fact the point; otherwise this argument would be swallowed without parsing if the "or source" part is executed.